Signal/src/ViewControllers/MediaPageViewController.swift

```diff
@@ -0,0 +1,706 @@
+        // 1. The media view ultimately lives in a zoomable scrollView. Getting both original positioning and the final positioning
+        //    correct, involves manipulating the zoomScale and position simultaneously, which results in non-linear movement,
+        //    especially noticeable on high resolution images.
+        // 2. For Video views, the AVPlayerLayer content does not scale with the presentation animation. So you instead get a full scale
+        //    video, wherein only the cropping is animated.
+        // Using a simple image view allows us to address both these problems relatively easily.
+        self.view.alpha = 0.0
+
+        self.pagerScrollView.isHidden = true
+        self.presentationView.isHidden = false
+        self.presentationView.layer.cornerRadius = OWSMessageCellCornerRadius
+
+        fromViewController.present(navController, animated: false) {
+
+            // 1. Fade in the entire view.
+            UIView.animate(withDuration: 0.1) {
+                self.replacingView?.alpha = 0.0
+                self.view.alpha = 1.0
+            }
+
+            self.presentationView.superview?.layoutIfNeeded()
+            self.applyFinalMediaViewConstraints()
+
+            // 2. Animate imageView from it's initial position, which should match where it was
+            // in the presenting view to it's final position, front and center in this view. This
+            // animation duration intentionally overlaps the previous
+            UIView.animate(withDuration: 0.2,
+                delay: 0.08,
+                options: .curveEaseOut,
+                animations: {
+
+                self.presentationView.layer.cornerRadius = 0
+                self.presentationView.superview?.layoutIfNeeded()
+
+                // We must lay out once *before* we centerMediaViewConstraints
+                // because it uses the imageView.frame to build the constraints
+                // that will center the imageView, and then once again *after*
+                // to ensure that the centered constraints are applied.
+                    //TODO
+//                [self centerMediaViewConstraints];
+//                [self.mediaView.superview layoutIfNeeded];
+                    self.view.backgroundColor = UIColor.white
+                },
+                completion: { (_: Bool) in
+                    // HACK: Setting the frame to itself *seems* like it should be a no-op, but
+                    // it ensures the content is drawn at the right frame. In particular I was
+                    // reproducibly seeing some images squished (they were EXIF rotated, maybe
+                    // related). similar to this report:
+                    // https://stackoverflow.com/questions/27961884/swift-uiimageview-stretched-aspect
+                    // TODO
+//                    self.mediaView.frame = self.mediaView.frame
```

Comment:
should this be uncommented per the comment above?

Reply:
Fixed, thanks!